controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java

```diff
@@ -76,6 +80,106 @@
     private static final OpStatsLogger DELETE_STREAM = STATS_LOGGER.createStats(MetricsNames.DELETE_STREAM);
     private final static String RESOURCE_PART_SEPARATOR = "_%_";
 
+    @Data
+    /**
+     * This class represents a 96 bit number with 32 bit msb encoded as integer and 64 bit lsb encoded as long.
+     * It is unsigned and only allows for non negative values. It also implements comparable interface and comparison involves
+     * first compariging msbs and if msbs are equal then we compare lsbs.
+     */
+    protected static class BigLong implements Comparable {
```

Comment:
I like this class. Can we:
1. Make it a top-level class?
2. I doubt there's a need for AtomicBigLong - see my comments below.
3. Can we make it more useful by having MSB and LSB both Long; for your use cases, you can just extract the last 32 bits of MSB. I think this may also solve the sign issues I brought up below.

I could see this possibly be reused in other cases - not ATM, though.

Reply:
> Make it a top-level class?
sure

>I doubt there's a need for AtomicBigLong - see my comments below.
AtomicBigLong provides us semantics of atomic increment and get operations. 
AtomicReference will be more costly with retries to achieve the same effect. 

> Can we make it more useful by having MSB and LSB both Long; for your use cases, you can just extract the last 32 bits of MSB. I think this may also solve the sign issues I brought up below.
sure